src/mono/mono/metadata/metadata.c

```diff
@@ -1495,6 +1495,114 @@ mono_metadata_decode_row_col_raw (const MonoTableInfo *t, int idx, guint col)
 	return 0;
 }
 
+static MONO_ALWAYS_INLINE guint32
+decode_value_scalar (const guint8 *ptr, const guint8 **new_ptr)
+{
+	guint32 result;
+	guint8 b = *ptr;
+
+	if ((b & 0x80) == 0){
+		result = b;
+		++ptr;
+	} else if ((b & 0x40) == 0){
+		result = ((b & 0x3f) << 8 | ptr [1]);
+		ptr += 2;
+	} else if (b != 0xff) {
+		result = ((b & 0x1f) << 24) |
+			(ptr [1] << 16) |
+			(ptr [2] << 8) |
+			ptr [3];
+		ptr += 4;
+	}
+	else {
+		result = (ptr [1] << 24) | (ptr [2] << 16) | (ptr [3] << 8) | ptr [4];
+		ptr += 5;
+	}
+
+	return result;
+}
+
+MONO_ALWAYS_INLINE guint32
+mono_metadata_decode_value_simd (const guint8 *ptr, const guint8 **new_ptr)
+{
+#ifdef __clang__
+	guint32 result;
+
+	typedef guint8 v64_u1 __attribute__ ((vector_size (8)));
+	typedef guint32 v64_u4 __attribute__ ((vector_size (8)));
+
+	// this will generate vectorized code on x64 and wasm as long as SIMD is enabled
+	// if simd isn't enabled, it generates fairly adequate scalar code.
+	// *(bytes *)ptr and *(guint32 *)ptr by themselves don't force an i32 load of
+	//  ptr in either x64 or wasm clang, so this is the only way to prefetch all the bytes
+	// without doing this, decode_value will do 5 conditional single-byte memory loads,
+	//  and each individual load is potentially bounds-checked. we produce one wide load
+	// we could overrun the source buffer by up to 7 bytes, but doing that on wasm is
+	//  safe unless we're decoding from the absolute end of memory.
+	// we pad all buffers by 16 bytes in mono_wasm_load_bytes_into_heap, so we're fine
+	// clang happily upsizes these 8-byte vectors into 16-byte ones for wasm, and uses 8-byte
+	//  insns on x64 as appropriate. armv8 looks fine too, albeit a little weird
+	union {
+		v64_u1 b;
+		v64_u4 i;
+	} v;
+	v.b = *(v64_u1 *)ptr;
```

Comment:
Do we need to `#ifdef NO_UNALIGNED_ACCESS` on this load?  How will this work on arm?

Reply:
Yeah, one thing I noticed is that this seems to generate aligned loads, and it shouldn't. I have to figure out how to fix that.